File: src/main/java/com/example/storage/service/impl/IssueServiceImpl.java
package com.example.storage.service.impl;

import com.example.storage.controller.dto.EquipmentDto;
import com.example.storage.controller.dto.IssueCreateDto;
import com.example.storage.entity.Employee;
import com.example.storage.entity.Equipment;
import com.example.storage.entity.EquipmentStatus;
import com.example.storage.entity.Issue;
import com.example.storage.exception.EmployeeNotFoundException;
import com.example.storage.exception.EquipmentNotAvailableException;
import com.example.storage.exception.EquipmentNotFoundException;
import com.example.storage.repository.EmployeeRepository;
import com.example.storage.repository.EquipmentRepository;
import com.example.storage.repository.IssueRepository;
import com.example.storage.service.EmployeeService;
import com.example.storage.service.IssueService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

@Service
@RequiredArgsConstructor
public class IssueServiceImpl implements IssueService {

    private final IssueRepository issueRepository;
    private final EmployeeRepository employeeRepository;
    private final EquipmentRepository equipmentRepository;
    private final EmployeeService employeeService;

    @Value("${issue.max-days}")
    private int maxDays;

    @Override
    @Transactional(readOnly = true)
    public Map<String, List<EquipmentDto>> getAvailableEquipment(String code) {
        employeeService.auth(code);

        Map<String, List<EquipmentDto>> result = new TreeMap<>();
        for (Equipment equipment : equipmentRepository.findAllByStatus(EquipmentStatus.AVAILABLE)) {
            String category = equipment.getCategory().getName();
            if (!result.containsKey(category)) {
                result.put(category, new java.util.ArrayList<>());
            }
            result.get(category).add(EquipmentDto.toDto(equipment));
        }

        return result;
    }

    @Override
    @Transactional
    public Issue create(String code, IssueCreateDto issueCreateDto) {
        LocalDate today = LocalDate.now(ZoneId.systemDefault());
        LocalDate returnDate = issueCreateDto.getReturnDate();
        if (returnDate.isBefore(today) || returnDate.isAfter(today.plusDays(maxDays))) {
            throw new IllegalArgumentException("Return date is out of range");
        }

        Employee employee = employeeRepository.findByCode(code)
                .orElseThrow(() -> new EmployeeNotFoundException("Employee with " + code + " code not found!"));

        long equipmentId = issueCreateDto.getEquipmentId();
        Equipment equipment = equipmentRepository.findById(equipmentId)
                .orElseThrow(() -> new EquipmentNotFoundException("Equipment with " + equipmentId + " id not found!"));

        if (equipment.getStatus() != EquipmentStatus.AVAILABLE) {
            throw new EquipmentNotAvailableException("Equipment " + equipment.getInventoryCode() + " is not available");
        }

        if (issueRepository.findByEquipmentAndReturnedAtIsNull(equipment).isPresent()) {
            throw new EquipmentNotAvailableException("Equipment " + equipment.getInventoryCode() + " is already issued");
        }

        equipment.setStatus(EquipmentStatus.ISSUED);
        equipmentRepository.save(equipment);

        Issue issue = Issue.builder()
                .issueDate(today)
                .returnDate(returnDate)
                .employee(employee)
                .equipment(equipment)
                .build();

        return issueRepository.save(issue);
    }
}

File: src/main/kotlin/ru/nto/storage/service/impl/IssueServiceImpl.kt
package ru.nto.storage.service.impl

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.nto.storage.controller.dto.EquipmentDto
import ru.nto.storage.controller.dto.IssueCreateDto
import ru.nto.storage.entity.EquipmentStatus
import ru.nto.storage.entity.Issue
import ru.nto.storage.exception.EmployeeNotFoundException
import ru.nto.storage.exception.EquipmentNotAvailableException
import ru.nto.storage.exception.EquipmentNotFoundException
import ru.nto.storage.repository.EmployeeRepository
import ru.nto.storage.repository.EquipmentRepository
import ru.nto.storage.repository.IssueRepository
import ru.nto.storage.service.EmployeeService
import ru.nto.storage.service.IssueService
import java.time.LocalDate
import java.time.ZoneId
import java.util.TreeMap

@Service
class IssueServiceImpl(
    private val issueRepository: IssueRepository,
    private val employeeRepository: EmployeeRepository,
    private val equipmentRepository: EquipmentRepository,
    private val employeeService: EmployeeService
) : IssueService {

    @Value("\${issue.max-days}")
    private var maxDays: Long = 0

    @Transactional(readOnly = true)
    override fun getAvailableEquipment(code: String): Map<String, List<EquipmentDto>> {
        employeeService.auth(code)

        val result = TreeMap<String, MutableList<EquipmentDto>>()
        for (equipment in equipmentRepository.findAllByStatus(EquipmentStatus.AVAILABLE)) {
            val category = equipment.category.name
            if (!result.containsKey(category)) {
                result[category] = mutableListOf()
            }
            result[category]!!.add(EquipmentDto.toDto(equipment))
        }

        return result
    }

    @Transactional
    override fun create(code: String, issueCreateDto: IssueCreateDto): Issue {
        val today = LocalDate.now(ZoneId.systemDefault())
        val returnDate = issueCreateDto.returnDate
        if (returnDate.isBefore(today) || returnDate.isAfter(today.plusDays(maxDays))) {
            throw IllegalArgumentException("Return date is out of range")
        }

        val employee = employeeRepository.findByCode(code)
            ?: throw EmployeeNotFoundException("Employee with $code code not found!")

        val equipmentId = issueCreateDto.equipmentId
        val equipment = equipmentRepository.findById(equipmentId).orElseThrow {
            EquipmentNotFoundException("Equipment with $equipmentId id not found!")
        }

        if (equipment.status != EquipmentStatus.AVAILABLE) {
            throw EquipmentNotAvailableException("Equipment ${equipment.inventoryCode} is not available")
        }

        if (issueRepository.findByEquipmentAndReturnedAtIsNull(equipment) != null) {
            throw EquipmentNotAvailableException("Equipment ${equipment.inventoryCode} is already issued")
        }

        equipment.status = EquipmentStatus.ISSUED
        equipmentRepository.save(equipment)

        val issue = Issue(
            issueDate = today,
            returnDate = returnDate,
            employee = employee,
            equipment = equipment
        )

        return issueRepository.save(issue)
    }
}
